File: algorithms/javaalgo/src/test/java/leetcode/number121to130/number121/BestTimeToBuyAndSellStockTest.java
package leetcode.number121to130.number121;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import org.junit.jupiter.api.Test;

public class BestTimeToBuyAndSellStockTest {

    @Test
    void test() {
        int[] prices1 = {7, 1, 5, 3, 6, 4};
        int[] prices2 = {2, 4, 1};

        assertAll(
            () -> assertThat(maxProfit(prices1)).isEqualTo(5),
            () -> assertThat(maxProfit(prices2)).isEqualTo(2)

        );
    }

    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int minPrice = Integer.MAX_VALUE;

        for (int price : prices) {
            minPrice = Math.min(minPrice, price);
            maxProfit = Math.max(maxProfit, price - minPrice);
        }
        return maxProfit;
    }
}

File: algorithms/javascriptalgo/leetcode/number121-130/number121/bestTimeToBuyAndSellStock.test.js
test("test", () => {
  const prices1 = [7, 1, 5, 3, 6, 4];
  const prices2 = [7, 6, 4, 3, 1];

  expect(maxProfit(prices1)).toEqual(5);
  expect(maxProfit(prices2)).toEqual(0);
  expect(maxProfit2(prices1)).toEqual(5);
  expect(maxProfit2(prices2)).toEqual(0);
});

function maxProfit(prices) {
  let answer = 0;
  let minPrice = Infinity;

  for (const price of prices) {
    minPrice = Math.min(minPrice, price);
    answer = Math.max(answer, price - minPrice);
  }

  return answer;
}

function maxProfit2(prices) {
  let buy = 0;
  let sell = 1;
  let answer = 0;

  while (sell < prices.length) {
    if (prices[buy] < prices[sell]) {
      let profit = prices[sell] - prices[buy];
      answer = Math.max(answer, profit);
    } else {
      buy = sell;
    }
    sell++;
  }

  return answer;
}
